streaming/src/test/scala/org/apache/spark/streaming/util/WriteAheadLogSuite.scala

```diff
@@ -190,283 +281,187 @@ class WriteAheadLogSuite extends SparkFunSuite with BeforeAndAfter {
+    assert(wal.getQueueLength === 0)
+    val event1 = "hello"
+    val event2 = "world"
+    val event3 = "this"
+    val event4 = "is"
+    val event5 = "doge"
+
+    eventFuture(wal, event1, 5L, numSuccess, numFail)
+    eventFuture(wal, event2, 10L, numSuccess, numFail)
+    eventFuture(wal, event3, 11L, numSuccess, numFail)
+    eventFuture(wal, event4, 12L, numSuccess, numFail)
+    eventFuture(wal, event5, 20L, numSuccess, numFail)
+
+    eventually(Eventually.timeout(2 seconds)) {
+      // the first element will immediately be taken and the rest will get queued
+      assert(wal.getQueueLength() == 4)
     }
-    if (closeLog) wal.close()
-    wal
-  }
-
-  /** Read data from a segments of a log file directly and return the list of byte buffers. */
-  def readDataManually(segments: Seq[FileBasedWriteAheadLogSegment]): Seq[String] = {
-    segments.map { segment =>
-      val reader = HdfsUtils.getInputStream(segment.path, hadoopConf)
-      try {
-        reader.seek(segment.offset)
-        val bytes = new Array[Byte](segment.length)
-        reader.readInt()
-        reader.readFully(bytes)
-        val data = Utils.deserialize[String](bytes)
-        reader.close()
-        data
-      } finally {
-        reader.close()
-      }
+    assert(numSuccess.get() === 0)
+    assert(numFail.get() === 0)
+    // remove block so that the writes are made
+    promise.success(null)
+
+    eventually(Eventually.timeout(2 seconds)) {
+      assert(wal.getQueueLength() == 0)
+      assert(numSuccess.get() === 5)
+      assert(numFail.get() == 0)
     }
   }
 
-  /** Read all the data from a log file directly and return the list of byte buffers. */
-  def readDataManually(file: String): Seq[String] = {
-    val reader = HdfsUtils.getInputStream(file, hadoopConf)
-    val buffer = new ArrayBuffer[String]
-    try {
-      while (true) {
-        // Read till EOF is thrown
-        val length = reader.readInt()
-        val bytes = new Array[Byte](length)
-        reader.read(bytes)
-        buffer += Utils.deserialize[String](bytes)
-      }
-    } catch {
-      case ex: EOFException =>
-    } finally {
-      reader.close()
+  test("BatchedWriteAheadLog - name log with aggregated entries with the timestamp of last entry") {
+    val wal = new BatchedWriteAheadLog(fileBasedWAL)
+    // block the write so that we can batch some records
+    val promise = writeBlockingPromise(fileBasedWAL)
+
+    val event1 = "hello"
+    val event2 = "world"
+    val event3 = "this"
+    val event4 = "is"
+    val event5 = "doge"
+
+    eventFuture(wal, event1, 3L) // 3 will automatically be flushed for the first write
+    // rest of the records will be batched while it takes 3 to get written
+    eventFuture(wal, event2, 5L)
+    eventFuture(wal, event3, 8L)
+    eventFuture(wal, event4, 12L)
+    eventFuture(wal, event5, 10L)
+    promise.success(true)
+
+    verify(fileBasedWAL, times(1)).write(any[ByteBuffer], meq(3L))
```

Comment:
This line should also be in an `eventually` block in case Jenkins is extremely slow.